### PR TITLE
[SPARK-9209] Using executor allocation, a executor is removed but it exists in ExecutorsPage of the web ui

### PR DESCRIPTION
**Pull request by KaiXinXiaoLei**
I set "spark.dynamicAllocation.enabled = true”, and  run a big job.  After some minutes, in driver, a executor is  asked to remove. Then it's removed successfully, and the process of this executor is not exist. But it exists in ExecutorsPage of the web ui.

The log in driver :
2015-07-17 11:48:14,543 | INFO  | [sparkDriver-akka.actor.default-dispatcher-3] | Removing block manager BlockManagerId(264, 172.1.1.8, 23811) 
2015-07-17 11:48:14,543 | INFO  | [dag-scheduler-event-loop] | Removed 264 successfully in removeExecutor 
2015-07-17 11:48:21,226 | INFO  | [sparkDriver-akka.actor.default-dispatcher-3] | Registering block manager 172.1.1.8:23811 with 10.4 GB RAM, BlockManagerId(264, 172.1.1.8, 23811) 
2015-07-17 11:48:21,228 | INFO  | [sparkDriver-akka.actor.default-dispatcher-3] | Added broadcast_781_piece0 in memory on 172.1.1.8:23811 (size: 38.6 KB, free: 10.4 GB)  
2015-07-17 11:48:35,277 | ERROR | [sparkDriver-akka.actor.default-dispatcher-16] | Lost executor 264 on datasight-195: remote Rpc client disassociated 
2015-07-17 11:48:35,277 | WARN  | [sparkDriver-akka.actor.default-dispatcher-4] | Association with remote system [akka.tcp://sparkExecutor@datasight-195:23929] has failed, address is now gated for [5000] ms. Reason is: [Disassociated].
2015-07-17 11:48:35,277 | INFO  | [sparkDriver-akka.actor.default-dispatcher-16] | Re-queueing tasks for 264 from TaskSet 415.0 
2015-07-17 11:48:35,804 | INFO  | [SparkListenerBus] | Existing executor 264 has been removed (new total is 10)

### CODE REVIEW

**Comment by KaiXinXiaoLei** [comment]
The log about this executor :
2015-07-17 11:48:20,762 | ERROR | [SIGTERM handler] | RECEIVED SIGNAL 15: SIGTERM
2015-07-17 11:48:20,993 | WARN  | [driver-heartbeater] | Told to re-register on heartbeat 
2015-07-17 11:48:20,993 | INFO  | [driver-heartbeater] | BlockManager re-registering with master
2015-07-17 11:48:21,052 | INFO  | [Thread-1] | Shutdown hook called 
2015-07-17 11:48:21,224 | INFO  | [Thread-1] | Shutdown hook called 
2015-07-17 11:48:21,184 | INFO  | [driver-heartbeater] | Trying to register BlockManager 
2015-07-17 11:48:21,227 | INFO  | [driver-heartbeater] | Registered BlockManager 
2015-07-17 11:48:21,227 | INFO  | [driver-heartbeater] | Reporting 16 blocks to the master.

File: core/src/main/scala/org/apache/spark/ui/exec/ExecutorsPage.scala [hunk 1]
```diff
@@ -58,7 +58,11 @@ private[ui] class ExecutorsPage(
     val diskUsed = storageStatusList.map(_.diskUsed).sum
     val execInfo = for (statusId <- 0 until storageStatusList.size) yield
       ExecutorsPage.getExecInfo(listener, statusId)
-    val execInfoSorted = execInfo.sortBy(_.id)
+    val execInfoSorted = execInfo.filter(x => {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/exec/ExecutorsPage.scala]: What is this doing? this is inefficient (creates regexes every time) and has strange syntax. Boolean or is not invoked as a method; extra parens.

File: core/src/main/scala/org/apache/spark/ui/exec/ExecutorsTab.scala [hunk 2]
```diff
@@ -68,6 +72,7 @@ class ExecutorsListener(storageStatusListener: StorageStatusListener) extends Sp
   override def onExecutorRemoved(
       executorRemoved: SparkListenerExecutorRemoved): Unit = synchronized {
     val eid = executorRemoved.executorId
+    executorIds = executorIds.filter(x => (x != eid))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/exec/ExecutorsTab.scala]: I can't make out how this relates to the problem you specified.  Do you mind backing up and explaining in the JIRA first? I would close this PR as it does not look like the right approach to anything like what you're describing

**Comment by andrewor14** [comment]
add to whitelist

**Comment by KaiXinXiaoLei** [comment]
I think my change code is not the reason of this failed  test , And in my machine, this test passed.

**Comment by srowen** [comment]
@KaiXinXiaoLei I think there are still some issues with your code, mostly that you're not explaining the change.

**Comment by KaiXinXiaoLei** [comment]
@srowen I am sorry not  to make clear. My problem is : a executor is removed successfully but  it exists in ExecutorsPage of the web ui.

From code, i find the executor id in ExecutorsPage is got from it's blockManagerId:
   val status = listener.storageStatusList(statusId)
    val execId = status.blockManagerId.executorId
    val hostPort = status.blockManagerId.hostPort

So  i suggest, add a variable to save active executor id, then reading this value of variable  to show active executor  in ExecutorsPage, 

For example, i add a variable "executorIds" in ExecutorsTab.scala  to save active executor id, and in ExecutorsPage.scala, filter active executor id through  "executorIds" .

File: core/src/main/scala/org/apache/spark/ui/exec/ExecutorsPage.scala (hunk 1 above)
Comment (KaiXinXiaoLei) [inline core/src/main/scala/org/apache/spark/ui/exec/ExecutorsPage.scala]: @srowen 
   val execInfoSorted = execInfo.filter(x => {
      val numPattern = "[0-9]+".r
     val noIntId = numPattern.findFirstIn(x.id).isEmpty
      listener.getExecutorIds.contains(x.id).||(noIntId)
    }).sortBy(_.id)

I intent to filter active executor id. In Executorspage, the type of  executor id is digital or character, eg
executorspage
So i use regexes .

**Comment by srowen** [comment]
I don't think it makes sense to add yet another piece of bookkeeping to patch over another problem. Update the existing data structures in the listener class. In any event I'm saying this change isn't OK from a style perspective.

... but I'm still not clear from your picture whether the executor is shown in an active state or not? is it not shown to be finished?

**Comment by KaiXinXiaoLei** [comment]
@srowen In my picture, i just want to say the type of executor id is digital or character, eg: the executor id 5 and "driver" in picture. 

Now,  do you understand my problem? If my way to fix this problem is not OK, can you give me a idea to fix this problem? Thanks.

**Comment by srowen** [comment]
No, that's not what I'm referring to. It is hacky to filter executors based on numeric-or-not name but happens to work now. The problem is that most lines of code in this PR have a problem -- you have a `var` where a `val` would do, if you use collection mutation methods; there's an unnecessary getter; you're using methods like `:+` with a `.`, unnecessary parens, missing whitespace, etc. 

I think it's best to close this PR and start over, since I don't think the approach is right either. surely the problem, if there is one, is farther upstream? why does the listener have `StorageStatus` for removed executors?

In the JIRA, can you propose any easy reproduction of this case?

**Comment by srowen** [comment]
Do you mind closing this PR? I don't think we're making progress in discussing it and I don't think this is a correct change

**Comment by KaiXinXiaoLei** [comment]
ok. i close this. and will find the better way

**Pull request closed without merging**